File: 503. 下一个更大元素 II/main.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int n = nums.length;
        int[] p = new int[2*n];
        for (int i = 0; i < n; i++) {
            p[i] = nums[i];
        }
        for (int i = n; i < 2 * n; i++) {
            p[i] = nums[i-n];
        }
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n + i; j++) {
                if (nums[i] < p[j]) {
                    nums[i] = p[j];
                    break;
                }
            }
            if (nums[i] == p[i]) {
                nums[i] = -1;
            }
        }
        return nums;
    }
}

File: 503. 下一个更大元素 II/main.cpp
// 维护单调递减栈，遍历所有元素，如果栈顶元素小于当前元素，那么栈顶元素的下一个最大值就是当前元素（单调递减的特性）
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        int n = nums.size();
        vector<int> ret(n, -1);
        stack<int> st;
        for (int i = 0; i < 2 * n - 1; i++) {
            while (!st.empty() && nums[st.top()] < nums[i%n]) {
                ret[st.top()] = nums[i%n];
                st.pop();
            }
            st.push(i % n);
        }
        return ret;
    }
};